### PR TITLE
gen, protos: 0.3, single cert

### LINKED ISSUE #190: Proposal: Specify only leaf cert in bundle
Background: The Sigstore bundle includes the verification material for a signature. For a signature generated with a key, it includes a public key fingerprint. When generated with a certificate, it currently can include either just the leaf certificate, or the leaf and its chain. The chain should only include intermediates, although there are bundles that were generated with roots too (resulting in [this comment](https://github.com/sigstore/protobuf-specs/blob/main/protos/sigstore_common.proto#L177). 

The [X509CertificateChain](https://github.com/sigstore/protobuf-specs/blob/main/protos/sigstore_common.proto#L172) message is currently used in two places: In the verification material for a bundle, or in the trust root. The comments on the message are mostly relevant to the former case.

As discussed in sig-clients, we would like to propose removing support for specifying a chain and require that the chain be distributed out of band or in the trust root. Including the chain in the bundle is useful for classical PKI cases where an intermediate is untrusted and only used for chain building. However the bundle wasn't designed with private deployments in mind. We do still want to support those cases, but want to be intentional in doing so.

One way we could this is to avoid breaking changes is to add an [additional field to the oneof in VerificationMaterial](https://github.com/sigstore/protobuf-specs/blob/main/protos/sigstore_bundle.proto#L60-L63), specifying a single certificate. Clients could then choose whether to support a BYO PKI case using `x509_certificate_chain`, or only support a single `certificate`. We could also do this in a breaking way and drop `x509_certificate_chain`.

Clients for thoughts: @loosebazooka @woodruffw @kommendorkapten @bdehamer @codysoyland

### PR DESCRIPTION
**Pull request by woodruffw**
This adds a new `oneof` variant to `VerificationMaterials.contents` that specifically encodes a single X.509 leaf certificate. This ratchets down (soft) invariants that were previously documented on the `x509_certificate_chain` member.

In the process, it also bumps the bundle version (`0.3`) and moves some documentation around.

Closes #190.

### CODE REVIEW

File: protos/sigstore_bundle.proto
Comment (woodruffw) [inline protos/sigstore_bundle.proto:91]: Flagging: I numbered this as `5`, since we didn't reserve any additional `oneof` variants here. I don't understand protobuf well enough to be certain, but I *think* this preserves the wire format on older bundles 🙂

File: protos/sigstore_common.proto
Comment (woodruffw) [inline protos/sigstore_common.proto:175]: Flagging: I added this `NOTE` based on the discussion we had in the clients call -- it might be overkill, though.

File: protos/sigstore_bundle.proto
Comment (haydentherapper) [reply protos/sigstore_bundle.proto:91]: Yep, 5 is correct https://protobuf.dev/programming-guides/proto3/#oneof:~:text=Field%20numbers%20for%20oneof%20fields%20must%20be%20unique%20within%20the%20enclosing%20message.

File: protos/sigstore_bundle.proto
Comment (haydentherapper) [inline protos/sigstore_bundle.proto:35]: Just pressure testing - Do we want a new version at this point? I think yes. We could add the field without a new version to avoid this being a breaking change, but then the field would have to be optional.

File: protos/sigstore_bundle.proto
Comment (haydentherapper) [inline protos/sigstore_bundle.proto:76]: Should we define "handle" more precisely? Something about ignoring root CA certificates if present? As for intermediates, that's currently up to the client. Cosign is the outlier in that it'll read in any intermediates provided. Maybe we mention something about clients providing a separate path for PGI vs private instances?

File: protos/sigstore_bundle.proto
Comment (haydentherapper) [inline protos/sigstore_bundle.proto:82]: Form 1 should be allowed for self-managed keys. You can use the public instance of Rekor without the public instance of Fulcio. I'm assuming this comment is about using Fulcio? Is there a way we can convey that if so?

File: protos/sigstore_common.proto
Comment (haydentherapper) [reply protos/sigstore_common.proto:175]: It's not just used for path construction though, also for providing a root CA certificate for TUF use cases.

Maybe add proposed different use cases?

File: protos/sigstore_common.proto
Comment (haydentherapper) [inline protos/sigstore_common.proto:181]: It's only a comment so we can't enforce easily, but should we be more prescriptive regarding order? Something like "index n+1 SHOULD be the issuing certificate for index n"?

File: protos/sigstore_bundle.proto
Comment (woodruffw) [reply protos/sigstore_bundle.proto:35]: Yep, I think so as well -- IMO making the field non-optional would be confusing in context, since the other two `content` variants are marked as required.

(But maybe there are other good reasons to not bump the version here? Other client maintainers should probably chime in 🙂)

File: protos/sigstore_bundle.proto
Comment (woodruffw) [reply protos/sigstore_bundle.proto:76]: Yeah, good catch -- I'll make it explicit that verifiers SHOULD handle these by ignoring any superfluous CA certificates (either by simply ignoring them entirely, or by ensuring that their chain building process treats them as untrusted for the purposes of verification). If I understand correctly, that'll be consistent with what `cosign` does 🙂

File: protos/sigstore_common.proto
Comment (woodruffw) [reply protos/sigstore_common.proto:181]: We could do that, although there may be (I really hope nobody does this, but there's nothing stopping it) cross-issuing PKI topologies like this one:

```
RCA_A -> ICA_A'
RCA_B -> ICA_A''
ICA_A -> leaf
```

(i.e. `ICA_A` is a single logical CA with two separate CA certs issued by different root CAs, either of which can be a valid issuer for `leaf`).

In this topology, `N+1` could be either `ICA_A'` or `ICA_A''`, and both could be listed in the "chain" for completeness purposes. So it's not clear which to order first 🙂 

(In practice, maybe this doesn't matter since `SHOULD` isn't `MUST`?)

File: protos/sigstore_bundle.proto
Comment (woodruffw) [reply protos/sigstore_bundle.proto:82]: Yeah, I'll clarify that this is about using Fulcio/"keyless" flows.

File: protos/sigstore_bundle.proto
Comment (kommendorkapten) [reply protos/sigstore_bundle.proto:35]: I would think that a breaking change is desired _if_ we want clients to actually adopt the usage of the new `certificate`. If not, clients would need to write the certificate into both `x509_certificate_chain` and `certificate`, and I'm not sure that's desirable? 

I think we can roll this out quite easy by describing the change more of a clarification, and as part of that moving the certificate to a new field and making it explicit. This would of course still require clients to discriminate on the version, and this is less optimal for verifiers that would need to be compliant with multiple versions.

File: protos/sigstore_common.proto
Comment (kommendorkapten) [reply protos/sigstore_common.proto:181]: I agree with @haydentherapper on being more prescriptive here. And as @woodruffw points out, if someone is doing a multi-rooted tree, they are are allowed as it's a `SHOULD`, and also, the verifier would probably be aware of this (i.e have both `RCA_A` and `RCA_B`) and should so be able to construct a chain regardless.

File: gen/jsonschema/schemas/Bundle.schema.json
Comment (loosebazooka) [inline gen/jsonschema/schemas/Bundle.schema.json:68]: would x509certificate be more consistent or is it just kind of redundant?

I guess this naming is in line with X509CertificateChain.certificates?

File: protos/sigstore_common.proto
Comment (woodruffw) [reply protos/sigstore_common.proto:181]: Sounds good, I'll make it more prescriptive!

File: gen/jsonschema/schemas/Bundle.schema.json
Comment (woodruffw) [reply gen/jsonschema/schemas/Bundle.schema.json:68]: Yeah, I was thinking the same. `x509_certificate` would be consistent with the other variant, but it feels a bit redundant. I have no strong opinions here 🙂

File: protos/sigstore_common.proto
Comment (loosebazooka) [reply protos/sigstore_common.proto:181]: Is there some situation in which we're expected to parse out this multirooted chain?

File: gen/jsonschema/schemas/Bundle.schema.json
Comment (loosebazooka) [reply gen/jsonschema/schemas/Bundle.schema.json:68]: just leave as is unless @haydentherapper or @kommendorkapten chime in?

File: protos/sigstore_common.proto
Comment (woodruffw) [reply protos/sigstore_common.proto:181]: 6d2a288 strengthens the prescription in the bundle case, PTAL 🙂

Comment (woodruffw) [reply protos/sigstore_common.proto:181]: > Is there some situation in which we're expected to parse out this multirooted chain?

This could show up in a BYO PKI scenario where the root of trust contains multiple separat trust anchors, but I think it's pretty unlikely -- I mostly produced it to illustrate that the chain "order" is ill-defined outside of basic PKIs.

File: protos/sigstore_bundle.proto
Comment (woodruffw) [reply protos/sigstore_bundle.proto:35]: > I would think that a breaking change is desired _if_ we want clients to actually adopt the usage of the new `certificate`. If not, clients would need to write the certificate into both `x509_certificate_chain` and `certificate`, and I'm not sure that's desirable?

My 0.02c would be for encouraging clients to adopt the `certificate` form -- the language in `VerificationMaterial` currently says that adopting the `0.3` bundle for keyless signing + PGI means that clients `MUST` use the `certificate` field instead of `x509_certificate_chain`. 

But yeah, this does make things a bit of a mess for verifying clients...

File: gen/jsonschema/schemas/Bundle.schema.json
Comment (haydentherapper) [reply gen/jsonschema/schemas/Bundle.schema.json:68]: What if we add support for SSH certificates? 😄 

No but really, leaving off `x509` is fine with me. We probably should have named the other field `certificate_chain` as x509 is a bit too in the weeds.

File: protos/sigstore_bundle.proto
Comment (haydentherapper) [reply protos/sigstore_bundle.proto:35]: For clients that wish to only support PGI, we could say just take the first cert of `cert-chain` or `certificate`.

Comment (kommendorkapten) [reply protos/sigstore_bundle.proto:35]: I think the current wording is good 👍 . My comment was a general about clients that would need to deal with multiple versions, not how the fields are used to day. I.e that regardless what we do, clients have to pay the price :) This is of course not specific to this spec, it's a general statement of concurrently supporting multiple versions.

File: protos/sigstore_common.proto
Comment (kommendorkapten) [inline protos/sigstore_common.proto:179]:
```suggestion
        // In some contexts (such as `VerificationMaterial.x509_certificate_chain`), this sequence
```

I believe this is what you meant? (Leaf is the first, then any intermediates.)

File: protos/sigstore_bundle.proto
Comment (woodruffw) [reply protos/sigstore_bundle.proto:35]: Yeah, further version proliferation isn't ideal.

In the interest of keeping the number of version iterations to a minimum, maybe we should roll this up with a few other changes that we expect to occur in a new version? In particular, I'm thinking of #189, #183, #66, and maybe some others.

File: protos/sigstore_common.proto
Comment (woodruffw) [reply protos/sigstore_common.proto:179]: Yes, thanks!

File: protos/sigstore_bundle.proto
Comment (kommendorkapten) [reply protos/sigstore_bundle.proto:35]: Yes, maybe it's time we plan for a next release, i.e bulk more things together as you mentioned. Don't think there is any urgency around getting `v0.3` out?

**Review by kommendorkapten — DISMISSED** [review]
Thanks for doing this 🚀

File: protos/sigstore_bundle.proto
Comment (woodruffw) [reply protos/sigstore_bundle.proto:35]: No extreme urgency, although I would definitely like to get #189 merged sooner rather than later 🙂 -- there's a lot of work that isn't strictly blocked on it, but will probably become a client interop challenge unless we eventually enumerate the various sigs/hashes that can occur on the PGI.

**Review by haydentherapper — APPROVED** [review]
Nice work!

**Comment by woodruffw** [comment]
In terms of procedure, should we merge here and let `0.3` live on `main`? I figure that's fine since it won't be tagged or released anywhere yet, but just confirming 🙂

**Review by kommendorkapten — APPROVED** [review]
Agree, merge this and let `0.3` live at `main` until we cut a release.

**Pull request merged**
